### DISCUSSION TITLE
How to use a pre defined offscreencanvas to create GL context?

### DISCUSSION THREAD

**Discussion by lzw1926** [emscripten-core/emscripten#23255] [Q&A] [answered]
Noticed that this `DISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR` option opt out Module['canvas'] mechanism as default, so seems I can't pass an offscreencanvas to initial `Module` and use to create GL context. So what's the correct way to do that?

**Comment by sbc100** [comment]
I think there are some docs here: https://emscripten.org/docs/api_reference/html5.h.html#registration-functions

**Reply by lzw1926** [reply]
> I think there are some docs here: https://emscripten.org/docs/api_reference/html5.h.html#registration-functions

Didn't get any on this question

**Reply by sbc100** [reply]
Do none of the ways listed in that work for for you?  How about that final "If the above are insufficient for you" part?

```
The target parameter is the ID of the HTML element to which the callback registration is to be applied. This field has the following special meanings:

- EMSCRIPTEN_EVENT_TARGET_WINDOW: The event listener is applied to the JavaScript window object.

- EMSCRIPTEN_EVENT_TARGET_DOCUMENT: The event listener is applied to the JavaScript document object.

- EMSCRIPTEN_EVENT_TARGET_SCREEN: The event listener is applied to the JavaScript window.screen object.

- 0 or NULL: If building with the option -sDISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR (default), NULL denotes an invalid element. If building with legacy option -sDISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR=0 (not recommended), a default element is chosen automatically based on the event type.

- #canvas: If building with legacy option -sDISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR=0 (not recommended), the event listener is applied to the Emscripten default WebGL canvas element. If building with the option -sDISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR (default), #canvas is interpreted as a CSS query selector: “the first element with CSS ID ‘canvas’”.

- Any other string: A CSS selector lookup is performed to the DOM with the passed string, and the event listener is applied to the first element that matches the query.

If the above are insufficient for you, you can add custom mappings in JavaScript using something like

specialHTMLTargets["!canvas"] = Module.canvas;
```

**Reply by lzw1926** [reply]
Appreciate for the quote reply. Since the point here is to utilize a pre defined `OffscreenCanvas` from JS to create GL context in cpp side, any css selector ways don't work, seems the final one shall do, but it still make me no sense why the default `DISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR` remove the utilization of Module['canvas']

**Reply by sbc100** [reply, marked as answer]
I think there is some rational in PR that deprecated the old behaviour: https://github.com/emscripten-core/emscripten/pull/7977.  

That `-sDISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR=0` settings does still exist if you want to keep using the old behaviour.   Although I suppose we would like to remove it one day I don't know any plans yet.

**Reply by sbc100** [reply]
From that PR I think this part is the reason why:

```
It changes html5.h API event target lookups to utilize document.querySelector() function instead of document.getElementById() function. querySelector() is much more flexible, since it allows one to use arbitrary CSS selectors, and not just DOM IDs. One can query DOM IDs with # prefix, i.e. document.querySelector('#canvas') will find DOM element with ID "canvas".
```

**Reply by lzw1926** [reply]
Thank you and I'll go through the PR to get more contexts.
